yapapi/golem.py

```diff
@@ -374,18 +375,46 @@ async def main():
 
         cluster = Cluster(
             engine=self._engine,
-            service_class=service_class,
             payload=payload,
             expiration=expiration,
             respawn_unstarted_instances=respawn_unstarted_instances,
             network=network,
         )
-
         await self._engine.add_to_async_context(cluster)
-        cluster.spawn_instances(num_instances, instance_params, network_addresses)
+
+        instance_param_gen = self._resolve_instance_params(num_instances, instance_params)
+        for ix, single_instance_params in enumerate(instance_param_gen):
+            network_address = None
+            if network_addresses is not None and len(network_addresses) > ix:
+                network_address = network_addresses[ix]
+
+            service = service_class(**single_instance_params)  # type: ignore
+            cluster.add_instance(service, network_address)
 
         return cluster
 
+    def _resolve_instance_params(
```

Comment:
given that I would very much like to preserve the `spawn_instances` functionality (that is, adding multiple instances to the Cluster using the current signature of `spawn_instances` (num_instances + params + network_addresses), I believe this function belongs to the Cluster as it was originally implemented

Reply:
I'll answer in the main comment section.